docs/design.md

```diff
@@ -0,0 +1,431 @@
+```
+
+
+##### schema: metadata/Control/v1
+
+This schema is the same as the `Document` schema, except it omits the
+`storagePolicy`, `layeringDefinition`, and `substitutions` keys, as these
+actions are not supported on `Control` documents.
+
+The complete list of valid `Control` document kinds is specified below along
+with descriptions of each document kind.
+
+##### schema: metadata/Tombstone/v1
+
+The only valid key in a `Tombstone` metadata section is `name`.  Additionally,
+the top-level `data` section should be omitted.
+
+### Layering
+
+When queried, Deckhand assembles the chain of partial documents in the order
+specified in the `deckhand/LayeringPolicy/v1` document. Each chain of partial
+documents is made up only of documents using a single `schema`.
+
+<!-- TODO: describe document chain construction w/ diagrams -->
+
+During each layering step, the list of `actions` will be applied in order.
+Supported actions are:
+
+* `merge` - a "deep" merge that layers new and modified data onto existing data
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
+<!-- TODO: Add examples and possibly figures -->
+
+### Substitution
+
+Concrete documents can be used as a source of substitution into other
+documents. This substitution is layer-independent.
+
+This is primarily designed as a mechanism for inserting secrets into
+configuration documents, but works for unencrypted source documents as well.
+
+<!-- TODO: Add example(s) maybe simple + an example with source layering? -->
+
+### Control Documents
+
+Control documents (documents which have `metadata.schema=metadata/Control/v1`),
+are special, and are used to control the behavior of Deckhand at runtime.  Only
+the following types control documents are allowed.
+
+#### DataSchema
+
+`DataSchema` documents are used by various services to register new schemas
+that Deckhand can use for validation. No `DataSchema` documents with names
+beginning with `deckhand/` or `metadata/` are allowed.
+
+<!-- TODO: give valid, tiny schema as example -->
+
+```yaml
+---
+schema: deckhand/DataSchema/v1
+metadata:
+  schema: metadata/Control/v1
+  name: promenade/Node/v1  # Specifies the documents to be used for validation.
+  labels:
+    application: promenade
+data:  # Valid JSON Schema is expected here.
+  $schema: http://blah
+```
+
+#### LayeringPolicy
+
+Only one `LayeringPolicy` document can exist within the system at any time.
+It is an error to attempt to insert a new `LayeringPolicy` document if it has
+a different `metadata.name` than the existing document. If the names match,
+it is treated as an update to the existing document.
+
+This document defines the strict order in which documents are merged together
+from their component parts. It should result in a validation error if a
+document refers to a layer not specified in the `LayeringPolicy`.
```

Comment:
1. I think that it is reasonable to issue warnings on the POST, and let the client ignore them all they want. 
200 OK
{ ignorable_warnings: [ ... ] }
(or whatever)

The service should diligently store the document as given. 
The stricter validations later will take care of things if they don't get settled by subsequent POSTs.

Dry run posts seem like feature creep, and would beg the question of wanting to have a whole series of POSTs run together as a dry run - definitely doable, but probably not high priority right now.  If we were going to do this, I think I'd rather allow for deletions of "unverified" versions,  maybe tied to some final validation that locks a version as permanent.